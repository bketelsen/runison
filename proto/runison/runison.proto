syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.routeguide";
option java_outer_classname = "RouteGuideProto";

package runison;

enum ChangeType {
  CLIENTADD = 0;
  CLIENTMODIFY = 1;
  CLIENTDELETE = 2;
  SERVERADD = 3;
  SERVERMODIFY = 4;
  SERVERDELETE = 5;
}
// Interface exported by the server.
service Synchronizer {
  // A simple RPC.
  //
  // Obtains the feature at a given position.
  //
  // A feature with an empty name is returned if there's no feature at the given
  // position.

  rpc GetChangeSet(Entries) returns (ChangeSetResponse) {}
}
message Node {
  // Node is a directory
  bool dir = 1;
  // Node is a file
  bool file = 2;
  // Node is a symlink
  bool symlink = 3;
  // File name
  string name = 4;
  // File path
  string path = 5;
  // File relative path
  string relative_path = 6;
  // File length
  uint64 len = 7;
  // Modified time seconds since Unix Epoch
  uint64 mod_seconds = 8;
  // Modified time nanoseconds since Unix Epoch
  uint32 mod_nano = 9;
  // Inode in filesystem
  uint64 inode = 10;
  // Base path of synchronization 
  string root_path = 11;
}

message Entries {
  // State of client
  map<string, Node> nodes = 1;
}
message ChangeSetResponse {
  repeated Change change = 1;
}
message Change {
  // Type of Change
  ChangeType change_type = 1;
  // Node that changed
  Node node = 2;
}
// Interface exported by the server.
service RouteGuide {
  // A simple RPC.
  //
  // Obtains the feature at a given position.
  //
  // A feature with an empty name is returned if there's no feature at the given
  // position.
  rpc GetFeature(Point) returns (Feature) {}

  // A server-to-client streaming RPC.
  //
  // Obtains the Features available within the given Rectangle.  Results are
  // streamed rather than returned at once (e.g. in a response message with a
  // repeated field), as the rectangle may cover a large area and contain a
  // huge number of features.
  rpc ListFeatures(Rectangle) returns (stream Feature) {}

  // A client-to-server streaming RPC.
  //
  // Accepts a stream of Points on a route being traversed, returning a
  // RouteSummary when traversal is completed.
  rpc RecordRoute(stream Point) returns (RouteSummary) {}

  // A Bidirectional streaming RPC.
  //
  // Accepts a stream of RouteNotes sent while a route is being traversed,
  // while receiving other RouteNotes (e.g. from other users).
  rpc RouteChat(stream RouteNote) returns (stream RouteNote) {}
}



// Points are represented as latitude-longitude pairs in the E7 representation
// (degrees multiplied by 10**7 and rounded to the nearest integer).
// Latitudes should be in the range +/- 90 degrees and longitude should be in
// the range +/- 180 degrees (inclusive).
message Point {
  int32 latitude = 1;
  int32 longitude = 2;
}

// A latitude-longitude rectangle, represented as two diagonally opposite
// points "lo" and "hi".
message Rectangle {
  // One corner of the rectangle.
  Point lo = 1;

  // The other corner of the rectangle.
  Point hi = 2;
}

// A feature names something at a given point.
//
// If a feature could not be named, the name is empty.
message Feature {
  // The name of the feature.
  string name = 1;

  // The point where the feature is detected.
  Point location = 2;
}

// A RouteNote is a message sent while at a given point.
message RouteNote {
  // The location from which the message is sent.
  Point location = 1;

  // The message to be sent.
  string message = 2;
}

// A RouteSummary is received in response to a RecordRoute rpc.
//
// It contains the number of individual points received, the number of
// detected features, and the total distance covered as the cumulative sum of
// the distance between each point.
message RouteSummary {
  // The number of points received.
  int32 point_count = 1;

  // The number of known features passed while traversing the route.
  int32 feature_count = 2;

  // The distance covered in metres.
  int32 distance = 3;

  // The duration of the traversal in seconds.
  int32 elapsed_time = 4;
}
