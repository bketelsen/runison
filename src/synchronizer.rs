use std::{
    collections::BTreeMap,
    fs,
    io::{self, BufReader},
    path::PathBuf,
    time::Instant,
};

use glob::Pattern;
use indicatif::{HumanDuration, ProgressBar, ProgressStyle};
use serde::{Deserialize, Serialize};
use walkdir::{DirEntry, WalkDir};

use crate::{config::Config, node::Node};

#[derive(Debug, Serialize, Deserialize, PartialEq, Copy, Clone)]
// Operational status of the process
pub enum Status {
    Starting,
    Indexing,
    Running,
    Stopping,
}
#[derive(Debug, Serialize, Deserialize, PartialEq, Copy, Clone)]
pub enum ChangeType {
    Added,
    Modified,
    Deleted,
}
#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]

pub struct Change {
    pub change_type: ChangeType,
    pub node: Node,
}

pub struct Synchronizer {
    pub entries: BTreeMap<String, Node>,
    config: Config,
    pub first_run: bool,
}
impl Synchronizer {
    pub fn new(config: Config) -> Option<Synchronizer> {
        Some(Synchronizer {
            entries: BTreeMap::new(),
            config,
            first_run: false,
        })
    }
    fn move_index(&mut self) -> io::Result<()> {
        // skip the move if the file won't be there
        if self.first_run {
            return Ok(());
        }
        let config = self.config.clone();
        let rp = String::from(config.root.path.clone());

        let mut archive = PathBuf::from(&rp);
        archive.push(".runison-current");

        let mut newarchive = PathBuf::from(&rp);
        newarchive.push(".runison-previous");
        println!("Moving index to {:?}", newarchive.display());
        fs::rename(archive, newarchive) // Rename a.txt to b.txt
    }
    pub fn index(&mut self) {
        let started = Instant::now();
        match self.move_index() {
            Ok(_) => {}
            Err(_) => {
                self.first_run = true;
            }
        }
        println!("Indexing files...");
        let pb = ProgressBar::new_spinner();
        pb.enable_steady_tick(200);
        pb.set_style(
            ProgressStyle::default_spinner()
                .tick_chars("/|\\- ")
                .template("{spinner:.dim.bold} indexing: {wide_msg}"),
        );
        let config = self.config.clone();
        let rp = String::from(config.root.path.clone());
        //    self.entries
        //       .insert(".".to_string(), Node::from_path(&rp, &config).unwrap());
        for entry in WalkDir::new(&rp)
            .into_iter()
            .filter_entry(|e| !ignored(e, &config.clone()))
        {
            let root_path = PathBuf::from(&rp);
            match entry {
                Ok(ent) => {
                    let config = self.config.clone();
                    let fp: String;

                    // if the path of the entry is the same as
                    // the root path, the entry key will be "" unless
                    // we specify it manually
                    if ent.path().to_str().unwrap().to_string().len()
                        == root_path.to_str().unwrap().to_string().len()
                    {
                        fp = root_path.to_str().unwrap().to_string();
                    } else {
                        fp = ent
                            .path()
                            .strip_prefix(&rp)
                            .unwrap()
                            .to_str()
                            .unwrap()
                            .to_string();
                    }
                    pb.set_message(&fp.clone());
                    self.entries
                        .insert(fp.clone(), Node::from_path(&fp, &config).unwrap());
                    pb.tick();
                }
                Err(_) => {}
            }
        }
        pb.finish_and_clear();
        println!("Done indexing in {}", HumanDuration(started.elapsed()));

        let mut archive = PathBuf::from(&rp);
        archive.push(".runison-current");
        println!("{:?}", archive.display());
        {
            let f = std::fs::File::create(archive).unwrap();
            bincode::serialize_into(f, &self.entries).unwrap();
        }
    }
    pub fn remote_changes(&mut self, tree: BTreeMap<String, Node>) -> Option<Vec<Change>> {
        println!("Detecting file changeset...");

        let started = Instant::now();
        let mut changes = Vec::new();
        for (path, node) in &self.entries {
            if let Some(remote) = tree.get(path) {
                // exists in both, check for change
                if !node.is_dir && node.modified != remote.modified {
                    // changed file
                    changes.push(Change {
                        change_type: ChangeType::Modified,
                        node: node.clone(),
                    })
                }
            } else {
                // doesn't exist in previous, is new file
                changes.push(Change {
                    change_type: ChangeType::Added,
                    node: node.clone(),
                })
            }
        }
        for (path, local) in tree {
            match self.entries.get(&path) {
                Some(_) => {}
                None => {
                    // current file is deleted

                    changes.push(Change {
                        change_type: ChangeType::Deleted,
                        node: local.clone(),
                    })
                }
            }
        }

        println!(
            "Done creating changeset in {}",
            HumanDuration(started.elapsed())
        );
        if changes.len() > 0 {
            return Some(changes);
        }

        None
    }
    pub fn local_changes(&mut self) -> Option<Vec<Change>> {
        println!("Detecting changed files...");
        if self.first_run {
            return None;
        }
        let started = Instant::now();
        let pb = ProgressBar::new_spinner();
        pb.enable_steady_tick(200);
        pb.set_style(
            ProgressStyle::default_spinner()
                .tick_chars("/|\\- ")
                .template("{spinner:.dim.bold} found: {wide_msg}"),
        );
        let config = self.config.clone();
        let rp = String::from(config.root.path.clone());

        let mut archive = PathBuf::from(&rp);
        archive.push(".runison-current");

        let mut oldarchive = PathBuf::from(&rp);
        oldarchive.push(".runison-previous");

        // TODO: If there is no current, then there is no local changeset.
        let current = std::fs::File::open(archive).unwrap();
        let reader = BufReader::new(current);

        let ca: std::result::Result<BTreeMap<String, Node>, Box<bincode::ErrorKind>> =
            bincode::deserialize_from(reader);
        match ca {
            Ok(current_archive) => {
                let previous = std::fs::File::open(oldarchive).unwrap();
                let preader = BufReader::new(previous);
                let pa: std::result::Result<BTreeMap<String, Node>, Box<bincode::ErrorKind>> =
                    bincode::deserialize_from(preader);
                match pa {
                    Ok(previous_archive) => {
                        let mut changes = Vec::new();
                        for (path, node) in &current_archive {
                            if let Some(prev) = previous_archive.get(path) {
                                // exists in both, check for change
                                if !node.is_dir && node.modified != prev.modified {
                                    // changed file
                                    pb.set_message(node.name.clone().to_str().unwrap());
                                    pb.tick();
                                    changes.push(Change {
                                        change_type: ChangeType::Modified,
                                        node: node.clone(),
                                    })
                                }
                            } else {
                                // doesn't exist in previous, is new file
                                pb.set_message(node.name.clone().to_str().unwrap());
                                pb.tick();
                                changes.push(Change {
                                    change_type: ChangeType::Added,
                                    node: node.clone(),
                                })
                            }
                        }
                        for (path, prev) in previous_archive {
                            match &current_archive.get(&path) {
                                Some(_) => {}
                                None => {
                                    // current file is deleted

                                    pb.set_message(prev.name.clone().to_str().unwrap());
                                    pb.tick();
                                    changes.push(Change {
                                        change_type: ChangeType::Deleted,
                                        node: prev.clone(),
                                    })
                                }
                            }
                        }

                        pb.finish_and_clear();
                        println!(
                            "Done scanning local changes in {}",
                            HumanDuration(started.elapsed())
                        );
                        return Some(changes);
                    }
                    Err(e) => {
                        println!("Error: {:?}", e);
                        return None;
                    }
                }
            }
            Err(e) => {
                println!("Error: {:?}", e);
                return None;
            }
        }
    }
}

// check ignored files and directories, returning true if
// the current entry should be ignored
pub fn ignored(entry: &DirEntry, config: &Config) -> bool {
    for pat in config.ignore.path.clone() {
        if Pattern::new(&pat)
            .unwrap()
            .matches(entry.path().to_str().unwrap())
        {
            return true;
        }
    }
    if Pattern::new(".runison-*")
        .unwrap()
        .matches(entry.path().file_name().unwrap().to_str().unwrap())
    {
        return true;
    }
    for pat in config.ignore.name.clone() {
        if Pattern::new(&pat)
            .unwrap()
            .matches(entry.path().file_name().unwrap().to_str().unwrap())
        {
            return true;
        }
    }
    false
}
