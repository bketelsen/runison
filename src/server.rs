use std::sync::Arc;

use prost_types::compiler::code_generator_response::File;
use tonic::transport::Server;
use tonic::{Request, Response, Status};

use runison::synchronizer_server::{Synchronizer, SynchronizerServer};

use runison::{ChangeSetResponse, Entries};
use std::path::PathBuf;
use structopt::StructOpt;

pub mod runison {
    tonic::include_proto!("runison");
}

mod config;
mod node;
mod synchronizer;
use synchronizer::Synchronizer as FileSynchronizer;

#[derive(Debug, StructOpt)]
#[structopt(name = "runison-server", about = "A modern file synchronization tool.")]
struct Opt {
    /// Activate debug mode
    // short and long flags (-d, --debug) will be deduced from the field's name
    #[structopt(short, long)]
    debug: bool,

    /// Configuration file
    #[structopt(short = "c", long = "config", parse(from_os_str))]
    config: PathBuf,
}
#[derive(Copy, Clone)]
struct SyncWrapper<'a> {
    sync_ref: &'a FileSynchronizer,
}
pub struct SynchronizerService<'a> {
    synchronizer: SyncWrapper<'a>,
}
#[tonic::async_trait]
impl Synchronizer for SynchronizerService {
    async fn get_change_set(
        &self,
        request: Request<Entries>,
    ) -> Result<Response<ChangeSetResponse>, Status> {
        println!("GetChangeSet = {:?}", request);
        &self
            .synchronizer
            .sync_ref
            .remote_changes(request.into_inner());
        Ok(Response::new(ChangeSetResponse::default()))
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let opt = Opt::from_args();
    println!("{:?}", opt);
    let result = config::get_config(opt.config);

    match result {
        Ok(config) => {
            let addr = "[::1]:10000".parse().unwrap();

            println!("Synchronizer listening on: {}", addr);

            let synchronizer = SynchronizerService {
                synchronizer: SyncWrapper { sync_ref: &synch },
            };

            let svc = SynchronizerServer::new(synchronizer);

            Server::builder().add_service(svc).serve(addr).await;
        }

        Err(_) => {}
    }
    Ok(())
}
