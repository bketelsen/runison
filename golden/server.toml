[root]
path = "/Users/bjk/src"

[path]
directories = [
    "src",
    "dotfiles"
]
[ignore]
name = [
    "*~",
    ".*~"
]
path = [
    "*node_modules*",
    "*target*"

]